### ISSUE TITLE
Required to install to pg_catalog

### ISSUE THREAD

**Issue by keithf4** [pgq/pgq#22]
Why is this extension (along with pgq_ext) requiring itself to be installed to pg_catalog? Pretty sure this is highly discouraged and having the schema line in the control file means there's no easy way for a typical user to change this.

Even the sql files generated are attempting to install it to the `pgq` schema. If a schema is required, maybe at least change it to this to be consistent with the sql files. Long term it would be best to use the @extschema@ macro to let users set their own schema.

**Comment by markokr** [comment]
It is a coding style decision - there is only one pgq module per db, under namespace `pgq.*` so db clients and other db modules can find it.

1) Why do you want to change the `pg_catalog`?  pgq does not install anything under it, it is just guarateed to exist schema to enforce that only one instance can be installed.

2) If you want to make the `pgq` part dynamic, then you need to configure all clients with the dynamic schema, which is annoying but possible.  But how do the other SQL modules find it? (`pgq_node`, `pgq_coop`, `londiste`)

**Comment by keithf4** [comment]
This seems a very odd coding choice and so far the only extension I've seen like this, including all of the contrib modules that upstream provides. It was confusing at first because when you do a `\dx` it shows that the extension exists in the `pg_catalog` schema but I see now all the objects exist in the `pgq` schema. Any other extension you'd install, the schema shown by `\dx` should be where those extension's objects are located, so if anything I'd think this would confuse other clients and db modules trying to construct properly schema qualified queries. Also looking it up in the catalog does the same
```
keith=# select extname, nspname from pg_extension e join pg_namespace n on e.extnamespace = n.oid where extname = 'pgq' 
;
 extname |  nspname   
---------+------------
 pgq     | pg_catalog
```
The above lookup is how I've often allowed for being able to find out what namespace an extension's objects are in so I can do a properly schema qualified query using those objects. Example in pg_partman looking up how to find where pg_jobmon extension is installed - https://github.com/pgpartman/pg_partman/blob/master/sql/functions/create_partition_time.sql#L101

It's not about making the `pgq` part dynamic or being able to change it later. As it is now, any user of the extension still has to schema qualify all their calls to the extension objects, but if they try and look up what that schema is, it shows them it's `pg_catalog` not `pgq`. 

Also not sure how this is providing any guarantee of a single install. An extension can only be installed once per database anyway. Trying to install it to another schema fails
```
test=# create schema partman;
CREATE SCHEMA
test=# create extension pg_partman schema partman;
CREATE EXTENSION
test=# create schema test;
CREATE SCHEMA
test=# create extension pg_partman schema test;
ERROR:  extension "pg_partman" already exists
```
And this does not stop it being installed on other databases in a single PG instance if that's what you're implying. pg_catalog is unique per database.

**Comment by keithf4** [comment]
I think I realized after I hit enter why you're doing this. If the target schema doesn't exist before you create the extension, then it fails. Changed the control to `schema = 'pgq`` and tested
```
keith=# create extension pgq schema pgq;
ERROR:  schema "pgq" does not exist
```
This is common for any other extension out there. However, if you define the schema you want the extension to be installed to, then it will automatically create that schema for you based on what you have defined in the control file. I adjusted the code in pgq to remove the `create schema` statement and changed the control file. It now installs it to the `pgq` schema and shows up properly in the catalogs for others to see
```
keith@dell-laptop:~/src/pgq$ git diff
diff --git a/pgq.control b/pgq.control
index 142a871..4ef1d91 100644
--- a/pgq.control
+++ b/pgq.control
@@ -3,5 +3,5 @@ comment = 'Generic queue for PostgreSQL'
 default_version = '3.5.1'
 relocatable = false
 superuser = true
-schema = 'pg_catalog'
+schema = 'pgq'
 
diff --git a/structure/tables.sql b/structure/tables.sql
index 6eccb74..12a17a7 100644
--- a/structure/tables.sql
+++ b/structure/tables.sql
@@ -26,7 +26,7 @@ set client_min_messages = 'warning';
 set default_with_oids = 'off';
 
 -- drop schema if exists pgq cascade;
-create schema pgq;
+--create schema pgq;
 
 -- ----------------------------------------------------------------------
 -- Table: pgq.consumer
```
```
test=# \dn
       List of schemas
  Name   |       Owner       
---------+-------------------
 partman | keith
 public  | pg_database_owner
 test    | keith
(3 rows)

test=# \dx
                               List of installed extensions
    Name    | Version |   Schema   |                     Description                      
------------+---------+------------+------------------------------------------------------
 pg_partman | 5.1.0   | partman    | Extension to manage partitioned tables by time or ID
 plpgsql    | 1.0     | pg_catalog | PL/pgSQL procedural language
(2 rows)

test=# create extension pgq;
CREATE EXTENSION
test=# \dx
                               List of installed extensions
    Name    | Version |   Schema   |                     Description                      
------------+---------+------------+------------------------------------------------------
 pg_partman | 5.1.0   | partman    | Extension to manage partitioned tables by time or ID
 pgq        | 3.5.1   | pgq        | Generic queue for PostgreSQL
 plpgsql    | 1.0     | pg_catalog | PL/pgSQL procedural language
(3 rows)
```

**Comment by markokr** [comment]
I've played around with it a bit and indeed, setting `schema='pgq'` mostly works, which is nice to know.

But there are 2 issues with it:
- The schema `pgq` would not belong to extension, it would be created and live outside of it. And it cannot be added to extension during install/upgrade with `ALTER EXTENSION` as Postgres explicitly disallows it: `ERROR:  cannot add schema "pgq" to extension "pgq" because the schema contains the extension` This seems to create confusing situation - who is responsible for managing grants for that schema?
- How to create robust upgrade procedure for it - the linked `pg_cron` issue already demonstrates the fragility of changing extension layout.

Although I agree that `\dx` would look nicer with proper schema, I don't see any other aspects that such change would improve for developers or tooling.  So unless I'm missing something obvious, I would prefer to keep status quo.

**Comment by keithf4** [comment]
How does making the schema an extension object make any difference in managing grants for the extension? Users still have to apply grants to allow access to the objects within it either way since you cannot predict what users will need that access. I can grant you the scenario of the someone dropping the schema also dropping the extension. But that seems a privilege granting scenario. No user that uses the extension needs CREATE privileges on that schema, so just don't allow that.

What upgrade scenarios are you seeing that you require it being installed to a pre-determinted schema? I've been maintaining pg_partman for 10+ years with over 80 update files using the `@extschema@` macro for all internal extension reference calls. I always make sure all object calls are always fully schema qualified. It looks like you're already doing that, so just swap out `pgq` for `@extschema@` in your code. 

I also have a dependency on another extension that also allows custom schema installs as well (`pg_jobmon`). I simply look up the schema that that extension is installed to and then schema qualify all calls from there. In the scenario here where the extension appears to be installed to `pg_catalog` such a simple lookup would not work because that is not where the extension objects exist. You'd have to look up the schema of each individual object to ensure you're getting the right one. 

I do not allow any extension I maintain to be relocatable after installation. I can definitely see how that would be a problem.

**Comment by markokr** [comment]
I was talking about upgrade procedure for `schema=pg_catalog` => `schema=pgq` change.

**Comment by keithf4** [comment]
Yes, this does seem to have backed the project into a difficult corner. Another reason being you cannot change the schema of something set to pg_catalog

https://dba.stackexchange.com/questions/107389/is-it-recommended-to-install-extensions-into-pg-catalog-schema/107406#107406

At least in the case of pg_cron, the change was more recent and users have already been put in a situation of not being able to upgrade from versions prior to 1.5 that forced the `pg_catalog` schema. In that case it could be reversed and the only people affected would be those that had installed more recent versions.

So either way it seems the fix is to uninstalled and reinstall the extension in this case. Perhaps look at it as a change for the next major version?

**Comment by keithf4** [comment]
> I still see it as a coding style issue. Imagene a range of extension types, from one extreme to other:
> 
>     1. Extension with couple schema-less functions, installed into first writable schema from `search_path`.  User can explicitly override the target schema.  Functions are very likely called without schema, instead expecting `search_path` to work. This is the primary use-case Postgres extension API was designed for.
> 
>     2. Extension with one fixed schema.
> 
>     3. Big extension/framework/app with multiple fixed schemas, schemas are owned and managed by extension.  It seems natural to use `pg_catalog` here.
> 
> 
> The question is - what schema should be assigned for [2]? Unlike [3], there are two options that seem to fit - `pg_catalog` or the fixed schema.
> 
> It seems using same style as for [3] is preferable for extension that clearly does not follow style of [1].

For 1 or 2, if you were talking about any other object here, would `pg_catalog` ever be the answer to where you'd put a third-party object? Ask the core team about putting any objects into `pg_catalog`, and I'm sure you know the answer you'd get. No, you're not putting anything in that actual schema, but you're still setting it to something that it is discouraged from third-party users from touching.

`pg_catalog` does not seem to be the answer to me.  Even in the case of #3, you could set a "main" schema of sorts using the schema option in the control file but still put your objects in whatever schema you want. `pg_catalog`

**Comment by markokr** [comment]
In that example, extensions will not put any objects into `pg_catalog`, we are talking about a placeholder schema for API designed for [1].

Extensions that actually put objects into `pg_catalog` are completely different topic, and indeed core team might want to have an opinion about that.

>  Even in the case of 3, you could set a "main" schema of sorts using the schema option in the control file but still put your objects in whatever schema you want.

But see the problem here - the "main" schema will not be owned by the extension, but others are. How is that useful?

**Comment by keithf4** [comment]
I agree it's not ideal that the schema is not owned by the extension. But it's not limiting anything as far as I can see either. This seems like it should be a bug raised to be fixed in core and not hacked around like this. Perhaps whatever schema is set in the control file should be owned by the extension if it didn't previously exist?

I've never run into any issues with the extensions I've maintained since 9.1 being able to be installed to a custom schema (outside of security issues that even core had to solve with things like pg_dump as well. schema qualify everything!).  Contrib extensions are fine in custom schemas. A massive project like PostGIS and all its sub-extensions are working perfectly fine with either custom or fixed schemas using the control file.

**Comment by keithf4** [comment]
Even in case 2 with something like pgq where you have a static schema, they'd have to look that up outside of the database catalog in the source code or do some other catalog lookup on each individual object to see what schema it's in. You talk about being confusing to developers and this is what I think is confusing since it's not visible where one would think it would be (`pg_catalog.pg_extension`).

The scenario of one extension installing to many schemas is valid, but I honestly haven't seen it myself. So speculating on something that I don't think either of us has seen as a valid use case doesn't seem to be useful. I'd honestly seriously question that sort of design and ask why they don't just make them different extensions with dependencies on each other. 

And just in general with core's view of "don't touch pg_catalog" I don't think it's a good idea.

**Comment by keithf4** [comment]
So been thinking over this the last week or so. I fully understand that changing this extension away from the `pg_catalog` usage would be extremely difficult and may not ever happen. I think I was more concerned with it becoming a trend after seeing it happen for the first time with pg_cron myself. I am also slightly concerned that upstream may do something that would disallow this and could break extensions that do it.

Going forward, I believe the best bet for extension design is choosing one of 2 options:

1. Setting the schema with the `schema` tag in the control file. 
2. Allowing the user to set the schema at install with the `SCHEMA` flag to `CREATE EXTENSION`

Either one of these methods would put the burden of security on the extension developer to fully schema qualify all object calls or properly set the search_path in the function definition. Either one of the above methods could use the `@extschema@` macro to schema qualify all internal extension objects. Note that neither of these options allow the extension to be relocatable and I've personally never found a need for that. This can be controlled with the `relocatable` flag in the control file.

If you happen to be going to PGConf.dev this year, we are having an Extensions summit to discuss things like this. Be happy to discuss it further there!
